net: latmon: latency monitor service and sample

ldts opened the pull request:
Supports the EVL/Xenomai4 gpio benchmarking tool as described in the [ Xenomai Latmus documentation](https://evlproject.org/core/benchmarks/#latmus-gpio-response-time).

The **sample** code for the **Latmon service** uses the FRDM_K64F J2 socket as follows:
- PIN_20: tx pulse to SUT
- PIN_18: rx ack from SUT

The client code running on the SUT shall monitor the falling edge of PIN_20.

Example usage from the latmus client running Xenomai4:
`  $ latmus -I gpiochip2,23,falling-edge -O gpiochip2,21 ....
`

Running snapshot:
-  Upper terminal: Zephyr Latmon sample
-  Lower terminal: Xenomai4  Latmus execution

![image](https://github.com/user-attachments/assets/67d55150-f251-4e9e-8235-ad430bfbbd91)


A sample histogram plot:
![image](https://github.com/user-attachments/assets/1ad91d88-d48a-42a2-976c-1bcaefdf4dee)


Running Sample:
![WhatsApp Image 2025-03-23 at 09 07 40](https://github.com/user-attachments/assets/f6c21af3-0829-41cf-8c3a-4075bdea6a67)

jukkar commented:
This sample uses networking APIs but does not seem to be a networking sample but testing gpio latency. Perhaps this should be located somewhere under `samples/drivers/` instead?

pdgendt commented on samples/net/latmon/README.rst:
While it is not a rule I can refer to, a lot of the rst files keep a maximum of 100 line length, which I do prefer.

pdgendt commented on samples/net/latmon/src/latmon.h:
```diff
@@ -0,0 +1,28 @@
+struct latmon_net_request {
+	uint32_t period_usecs;
+	uint32_t histogram_cells;
+} __attribute__((__packed__));
```
Use `__packed` instead.

pdgendt commented on samples/net/latmon/src/main.c:
```diff
@@ -0,0 +1,718 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2024 Philippe Gerum <rpm@xenomai.org>
+ * Copyright (c) 2024 Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>
```
2025?

pdgendt commented on samples/net/latmon/src/main.c:
```diff
@@ -0,0 +1,718 @@
+static const struct gpio_dt_spec led_wait =
+	GPIO_DT_SPEC_GET(DT_ALIAS(led1), gpios);
+static const struct gpio_dt_spec led_dhcp =
+	GPIO_DT_SPEC_GET(DT_ALIAS(led2), gpios);
```
Not all board have 3 leds, maybe use `GPIO_DT_SPEC_GET_OR` and make them optional.

ldts replied:
ok

pdgendt commented on samples/net/latmon/src/main.c:
```diff
@@ -0,0 +1,718 @@
+
+	LOG_ERR("timeout waiting for ack ");
+
+	if (!atomic_clear(&ack_wait))
+		LOG_ERR("unexpected value clearing ack_wait ");
```
Always put curly braces around `if` blocks.

```suggestion
	if (!atomic_clear(&ack_wait)) {
		LOG_ERR("unexpected value clearing ack_wait ");
	}
```

ldts replied:
ok

bjarki-andreasen commented on samples/net/latmon/dts/bindings/latency-monitor.yaml:
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2024 Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>
+# SPDX-License-Identifier: Apache-2.0
+
+description: GPIO latency monitor
+
+compatible: "latency-monitor"
+
+child-binding:
+  properties:
+    gpios:
+      type: phandle-array
+      required: true
+
+    pinctrl-0:
+      type: phandle
+      required: false
```
This is not a device, these properties do not go in the devicetree bindings. If you need to specify gpios to use for your sample, use the special node `zephyr,user` and get the properties using `DT_PATH(zephyr_user)`

bjarki-andreasen reviewed (changes requested):
I do not believe this qualifies as a sample. From what I see reading the code, its an application used to calculate the latency of GPIO propagation between two systems, using a timer. For testing embedded systems, latencies are usually measured using digital analyzers to isolate the device under test, to avoid the propagated delays this application is seemingly introducing.

Samples are meant to show how to use zephyr, focusing on a specific feature like RTIO. I do not think anyone seeing this code will learn anything new about zephyr.

Why not keep this application downstream with the project it is designed for? I do not see what the zephyr project gains from adding and maintaining this application.

ldts commented:
> I do not believe this qualifies as a sample. From what I see reading the code, its an application used to calculate the latency of GPIO propagation between two systems, using a timer. For testing embedded systems, latencies are usually measured using digital analyzers to isolate the device under test, to avoid the propagated delays this application is seemingly introducing.
> 
> Samples are meant to show how to use zephyr, focusing on a specific feature like RTIO. I do not think anyone seeing this code will learn anything new about zephyr.
> 
> Why not keep this application downstream with the project it is designed for? I do not see what the zephyr project gains from adding and maintaining this application.

As I tried to explain in the [README.rst](https://github.com/zephyrproject-rtos/zephyr/blob/1ab4863dba5f7777418385c8629f89200882722a/samples/net/latmon/README.rst), this program is designed for real-time Linux systems, assisting in the evaluation of their interrupt latencies. The client code, [latmus](https://evlproject.org/core/testing/#latmus-program), runs on the system under test (SUT) and has been a part of the Xenomai project for over 20 years, spanning multiple iterations. However, since it can run with minimal modifications on various systems, the goal is to extract it into a standalone project or potentially integrate it into the Linux kernel tree - maybe under linux/tools. The ultimate objective is to develop an affordable, open-source, hardware-backed solution for measuring interrupt latency in a clear and reliable manner.

I understand your concern but I was hoping we could have a discussion about this. In my opinion, given its simplicity, the code serves as a real-world example of pinctrl, GPIO, and networking usage. Additionally, it contributes to the Linux community by supporting discussions on real-time performance data. If it would be helpful, I’d be happy to be added as a maintainer for this sample code.

ldts commented:
@jukkar @bjarki-andreasen should I add it to samples/boards instead? It seems to focus more on demonstrating functionality and hardware features. I'd like to unblock this discussion

bjarki-andreasen commented:
> @jukkar @bjarki-andreasen should I add it to samples/boards instead? It seems to focus more on demonstrating functionality and hardware features. I'd like to unblock this discussion

Could you address the main alternative to adding this as a sample, which is to maintain it as a normal application within the project it is a part of? There are an effectively never ending number of applications built to run on zephyr, projects like https://github.com/jakkra/ZSWatch and https://github.com/zmkfirmware/zmk are rather popular ones. Should we add those applications to samples as well? One could argue they demonstrate how to use many parts of zephyr as well, but that does not make them samples...

jukkar commented:
I do not have big issues with this sample, we have other samples that do somewhat similar things like zperf in networking. If this PR is accepted, then I suggest it is not placed under `samples/net` as it is using networking APIs but is not demonstrating some specific networking feature (we have various BSD socket samples in net already which do that).

bjarki-andreasen commented:
> @bjarki-andreasen I believe I have provided some sort of justification to what you are asking; I have provided links and data backing my request and offered my support as a maintainer.
> 
> You are pointing me to some neat repositories though quite frankly I fail to see their industry relevance beyond their own individuality - the problem they are trying to solve doesn't extend beyond their single use case. I also doubt they will be active in 20 years from now. I have pointed you to a project/framework that is a quarter of a century old and supporting many industrial applications (and companies).
> 
> The latmus tool - used to measure performance deviation - will help a wide range of products and markets. So with that in mind, I am still not sure why this Zephyr code can not be integrated as sample code. Is as generic as any, but it will help many.

Yes, "some sort of justification" has been provided, I do not find said justification sufficient.

> Now, please, could you help me understand the criteria to merge code to sample/boards? what does an application need to do be be accepted there?

`samples/boards` is for board specific samples, demoing board/vendor specific features like vendor specific device driver APIs. Its not relevant so I did not see a reason to expand upon it.

@jukkar Just like this sample does not demoing any specific networking feature, its not demoing any specific zephyr feature. Its not a sample... The "sample" in this PR is an endpoint/client for a completely independent project, it has no utility for zephyr.

@kartben What do you think? Is this a sample in your opinion? Should it be part of the zephyr codebase somewhere in any case?

bjarki-andreasen commented:
Have you considered contributing latmon as a network protocol, similar to how protocols like ptp are implemented? Then the sample could demo the usage of said protocol, which would be a zephyr feature :) The protocol being part of the sample is what I believe makes this an application, rather than demoing a zephyr feature.

something like:
```
typedef void (*net_latmon_ack_handler)(void);
int net_latmon_start(net_latmon_ack_handler ack_handler);
int net_latmon_send_data(const struct net_latmon_data *data);
int net_latmon_send_data(const struct net_latmon_data *data);
```
in
`include/zephyr/net/latmon.h`
which could be included with a kconfig like `config NET_LATMON`.

The sample would then be a demo of how to use the protocol.

ldts commented:
> Have you considered contributing latmon as a network protocol, similar to how protocols like ptp are implemented? Then the sample could demo the usage of said protocol, which would be a zephyr feature :) The protocol being part of the sample is what I believe makes this an application, rather than demoing a zephyr feature.
> 
> something like:
> 
> ```
> typedef void (*net_latmon_ack_handler)(void);
> int net_latmon_start(net_latmon_ack_handler ack_handler);
> int net_latmon_send_data(const struct net_latmon_data *data);
> ```
> 
> in `include/zephyr/net/latmon.h` which could be included with a kconfig like `config NET_LATMON`.
> 
> The sample would then be a demo of how to use the protocol.

ah, that is a neat idea. ok let me have a look and thanks a lot for pointing it out!

ldts commented:
keep alive.

ldts committed:
net: latmon: Add latency monitor service

Latmon service to interface with Xenomai's Latmus.

Link: https://evlproject.org/core/benchmarks/#latmus-gpio-response-time

Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>

ldts commented:
@bjarki-andreasen I still have to address issues from the original code so please do not dive into a formal review yet - could you however comment on the partitioning as a network service? is this what you had in mind when you suggested it?

bjarki-andreasen commented:
> @bjarki-andreasen I still have to address issues from the original code so please do not dive into a formal review yet - could you however comment on the partitioning as a network service? is this what you had in mind when you suggested it?
> 
> 

Yes, this looks very much like what I envisioned :)

ldts commented:
@bjarki-andreasen ready for review and thanks again for the suggestion.

bjarki-andreasen commented on subsys/net/lib/latmon/latmon.c:
```diff
@@ -0,0 +1,458 @@
+
+	LOG_INF("Waiting for connection...");
+
+	ret = poll(fds, 1, 5000);
```
Please provide the timeout as a parameter to the `net_latmon_connect` API rather than a "magic" value, unless 5000 is defined by the protocol, in which case, please define it, something like `#define NET_LATMON_CONNECT_TIMEOUT 5000`

bjarki-andreasen commented on subsys/net/lib/latmon/latmon.c:
```diff
@@ -0,0 +1,458 @@
+/* Waits for connection from Latmus */
+int net_latmon_connect(int socket, struct in_addr *ip)
+{
+	struct pollfd fds[1] = { {.fd = socket, .events = POLLIN } };
```
please use `struct zsock_pollfd` and `zsock_poll()`

bjarki-andreasen commented on subsys/net/lib/latmon/Kconfig:
```diff
@@ -0,0 +1,25 @@
+	select NET_SOCKETS
+	select NET_MGMT
+	select NET_MGMT_EVENT
+	select NET_LOG
```
`NET_LOG` is no a dependency of this module, its selected by the application

bjarki-andreasen commented on subsys/net/lib/latmon/Kconfig:
```diff
@@ -0,0 +1,25 @@
+	select NET_MGMT_EVENT
+	select NET_LOG
+	select POSIX_API
+	select NET_SHELL
```
NET_SHELL is not a dependency of this module

bjarki-andreasen commented on subsys/net/lib/latmon/latmon.c:
```diff
@@ -0,0 +1,458 @@
+ */
+
+ #include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(latmon, LOG_LEVEL_DBG);
```
Should be `LOG_MODULE_REGISTER(latmon, CONFIG_LATMON_LOG_LEVEL);`

bjarki-andreasen commented on subsys/net/lib/latmon/latmon.c:
```diff
@@ -0,0 +1,458 @@
+ #include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(latmon, LOG_LEVEL_DBG);
+
+#include <stdio.h>
```
stdio does not seem to be used, should it be `string.h` maybe?

bjarki-andreasen commented on subsys/net/lib/latmon/latmon.c:
```diff
@@ -0,0 +1,458 @@
+
+	sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+	if (sock < 0) {
+		perror("socket");
+		exit(1);
```
Please avoid POSIX functions in kernel subsystems. Not entirely sure if this would even compile on a non posix target even with POSIX_API selected, or what it would do :) Just return -1 and propagate it to the caller of `net_latmon_connect()`

bjarki-andreasen commented on subsys/net/lib/latmon/latmon.c:
```diff
@@ -0,0 +1,458 @@
+
+	s = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+	if (s < 0) {
+		perror("socket");
+		exit(1);
```
same here, propagate error, don't crash the kernel or whatever exit(1) would do :)

bjarki-andreasen commented on subsys/net/lib/latmon/Kconfig:
```diff
@@ -0,0 +1,25 @@
+	bool "Latency monitoring support"
+	help
+	  This option enables the latency monitoring support for Zephyr
+	select NET_IPV4
+	select NET_ARP
+	select NET_TCP
+	select NET_UDP
+	select NET_DHCPV4
+	select NET_SOCKETS
+	select NET_MGMT
+	select NET_MGMT_EVENT
```
Summarized, only NET_SOCKETS needs to be selected here, and maybe add a "depends on NET_TCP || NET_UDP". Only strictly necessary options may be selected by the module.

bjarki-andreasen reviewed (changes requested):
Summary:
* Please refactor the module a bit to not be dependent on POSIX, by using zephyr native APIs and avoiding functions like `perror` and `exit`.
* Cut back on the selections made by the module's Kconfig. Kconfig is supposed to list and/or select required dependencies only.
* Please add a test suite which tests the module using `CONFIG_NET_TEST=y`, see https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/net/icmp for example

ldts commented:
thanks for all the details @bjarki-andreasen will post the fixes by the end of the week.

ldts commented:
@bjarki-andreasen I have updated the PR and approved from legal.

ldts commented on include/zephyr/net/latmon.h:
```diff
@@ -0,0 +1,93 @@
+ */
+
+/*
+ * Copyright (c) 2025 Jorge A. Ramirez Ortiz <jorge.ramirez@oss.qualcomm.com>
```
I'll own the copyright for this code even though I am using the company identity.

ldts commented on MAINTAINERS.yml:
```diff
@@ -2987,6 +2987,22 @@ Networking:
   tests:
     - sample.net.gptp
 
+"Networking: Latmon":
```
thanks - yes, I was not aware of the process. sounds good

ldts commented on include/zephyr/net/latmon.h:
```diff
@@ -0,0 +1,93 @@
+ *
+ * @return A valid socket descriptor on success, negative errno code on failure.
+ */
+int net_latmon_get_socket(void);
```
yes, good point

ldts commented on subsys/net/lib/latmon/latmon.c:
```diff
@@ -0,0 +1,554 @@
+	LOG_INF("\tmonitoring stopped");
+}
+
+static int broadcast_ip_address(struct in_addr *ip_addr)
```
yes, I'll rather let the ipv6 support as a future enhancement. need to touch a number of files and including Xenomai to support both

ldts commented:
@jukkar thanks. updated.

ldts commented on samples/net/latmon/src/main.c:
```diff
@@ -0,0 +1,288 @@
+
+	/* Generate a falling edge pulse to the DUT */
+	key = irq_lock();
+	if (gpio_pin_set_dt(&pulse, 0)) {
```
not needed

kartben reviewed (changes requested):
Some documentation comments, will prob have more when I'm in front of a computer. Thanks for this contribution!

ldts commented on include/zephyr/net/latmon.h:
```diff
@@ -0,0 +1,93 @@
+ * negative errno code on failure.
+ *
+ */
+int net_latmon_connect(int socket, struct in_addr *ip);
```
latmus does not support ipv4

ldts commented on samples/net/latmon/src/main.c:
```diff
@@ -0,0 +1,305 @@
+
+	LOG_INF("DHCPv4: binding...");
+	net_mgmt_init_event_callback(cb, dhcp_handler,
+				     NET_EVENT_IPV4_ADDR_ADD);
```
awesome. thanks

kartben commented on doc/connectivity/networking/api/latmon.rst:
```diff
@@ -0,0 +1,168 @@
+monitoring process.
+
+
+4. API Reference
```
Please include an API reference section similar to https://github.com/zephyrproject-rtos/zephyr/blob/e0dbba5bd3ff18103fa3adb9c6efe2e44fcf278b/doc/connectivity/networking/api/mqtt.rst?plain=1#L170 though - and referencing the Doxygen group in your code sample (relevant_api) will also ensure it shows up there -  see https://docs.zephyrproject.org/latest/connectivity/networking/api/mqtt.html#api-reference

kartben commented:
@ldts also, as a general note:

"Unless they applied the reviewer’s recommendation exactly, authors must not resolve and hide comments, they must let the initial reviewer do it. The Zephyr project does not require all comments to be resolved before merge"
 https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#:~:text=Unless%20they%20applied%20the%20reviewer%E2%80%99s%20recommendation%20exactly%2C%20authors%20must%20not%20resolve%20and%20hide%20comments%2C%20they%20must%20let%20the%20initial%20reviewer%20do%20it.%20The%20Zephyr%20project%20does%20not%20require%20all%20comments%20to%20be%20resolved%20before%20merge

ldts commented:
> @ldts also, as a general note:
> 
> "Unless they applied the reviewer’s recommendation exactly, authors must not resolve and hide comments, they must let the initial reviewer do it. The Zephyr project does not require all comments to be resolved before merge" https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#:~:text=Unless%20they%20applied%20the%20reviewer%E2%80%99s%20recommendation%20exactly%2C%20authors%20must%20not%20resolve%20and%20hide%20comments%2C%20they%20must%20let%20the%20initial%20reviewer%20do%20it.%20The%20Zephyr%20project%20does%20not%20require%20all%20comments%20to%20be%20resolved%20before%20merge

yeah I was wondering about that when I was cleaning up (btw on every resolve I did take the action suggested but thanks for pointing it out). 

still, cant github be configured to not offer users the `resolve` button?

ldts commented on doc/connectivity/networking/api/latmon.rst:
```diff
@@ -0,0 +1,160 @@
+3. Key Components
+*****************
+
+3.1 **Socket Management**
```
ok. I assume "Manuel" has nothing to do with Fawlty Towers :)  
![image](https://github.com/user-attachments/assets/863a0a6d-eb21-4e5a-ab8f-6615065e2f9b)

ldts replied:
:)

jukkar commented on subsys/net/lib/latmon/latmon.c:
```diff
@@ -0,0 +1,563 @@
+}
+
+/* Get a socket to listen to Latmus requests */
+int net_latmon_get_socket(struct sockaddr_in *connection_addr)
```
But here if we have `struct sockaddr *connection_addr`, then we do not need to tweak the API to support IPv6.
Then the memcpy() below could look like this
```
if (connection_addr != NULL) {
	memcpy(&addr, (struct sockaddr_in *)connection_addr, sizeof(addr));
}
```

jukkar commented on subsys/net/lib/latmon/latmon.c:
```diff
@@ -0,0 +1,563 @@
+}
+
+/* Waits for connection from Latmus */
+int net_latmon_connect(int socket, struct in_addr *ip)
```
And here if we change this to `net_latmon_connect(int socket, struct sockaddr *ip)`, then there is no need to change the API to support IPv6

jukkar commented on samples/net/latmon/src/main.c:
```diff
@@ -0,0 +1,305 @@
+	return ret;
+}
+
+static struct in_addr *get_dhcp_address(void)
```
Note that the bound IP address is sent with the BOUND event, so there should be no need to traverse the IP address list.

jukkar commented on samples/net/latmon/src/main.c:
```diff
@@ -0,0 +1,287 @@
+	struct net_if *iface = net_if_get_default();
+
+	LOG_INF("DHCPv4: binding...");
+	net_mgmt_init_event_callback(cb, dhcp_handler,
+				     NET_EVENT_IPV4_DHCP_BOUND);
+	net_mgmt_add_event_callback(cb);
+	net_dhcpv4_start(iface);
+	k_sem_take(&dhcp_done, K_FOREVER);
```
Instead of creating a semaphore and a handler, you could do here
```
struct net_if_dhcpv4 *dhcpv4;
size_t info_len;

ret = net_mgmt_event_wait(NET_EVENT_IPV4_DHCP_BOUND, NULL, NULL, &dhcpv4, &info_len, timeout);
```
which would wait until the desired event is returned. The bound IPv4 address is then found in the dhcpv4 struct.

ldts replied:
I modified it slightly to avoid a memcpy.

ldts commented:
Compliance checks: MISRA (via [sonarqubecloud](https://github.com/apps/sonarqubecloud)) appears to require an else clause for every if-else construct, even in cases—like this one—where the else is functionally unnecessary.
how shall I proceed?

ldts commented:
the SonarCloud Code analysis output is bogus as ip is validated before calling strlen

ldts commented:
@Jukkar anything else on these Pars?

ldts commented:
can we merge this please?

jukkar commented:
@ldts 
* We could try to overcome sonarcloud issue by using `strnlen` instead of `strlen`
* There are "changes requested" by two devs that need to be cleared
* I had wished that would could make the APIs more IP address agnostic. The server does not support IPv6, but perhaps it will in the future. Regardless of that, it is pain to add support for IPv6 in the future if the APIs need to be changed. Minimally we should mark the latmon support as experimental in `subsys/net/lib/latmon/Kconfig` to make it easier to change the APIs when needed if it is not possible to make proper APIs atm.

ldts commented:
hi @jukkar I did discuss with Philippe and really there is no intention to support ipv4 so I'd rather follow the option of marking it as experimental.
ok I have re-requested the pending reviews (AFAIK I did address all issues).

ldts commented:
as part of the re-review I'll update the strlen to strnlen (even though not needed in this case)

bjarki-andreasen commented on samples/net/latmon/prj.conf:
```diff
@@ -0,0 +1,28 @@
+
+CONFIG_POSIX_API=y
+
+# General config
+CONFIG_LOG=y
+CONFIG_GPIO=y
+CONFIG_REBOOT=y
+CONFIG_TICKLESS_KERNEL=y
+CONFIG_CONSOLE_SUBSYS=y
+CONFIG_CONSOLE_GETLINE=y
+CONFIG_NEWLIB_LIBC=y
+CONFIG_INIT_STACKS=y
```
I believe only `CONFIG_GPIO=y` and `CONFIG_LOG=y` need to be explicitly enabled. other options are platform specific, and should have reasonable defaults (hello_world sure does not select CONFIG_TICKLESS_KERNEL=y or CONFIG_REBOOT=y for example)

bjarki-andreasen replied:
also zephyr is moving to picolibc :)

ldts replied:
all right, testing. thanks!

bjarki-andreasen commented on samples/net/latmon/src/main.c:
```diff
@@ -0,0 +1,263 @@
+	k_sem_reset(&ack_event);
+
+	/* Generate a falling edge pulse to the DUT */
+	key = irq_lock();
```
applications should use kernel primitives, `irq_lock()` is just there for compatibility reasons, instead, `struct k_spinlock` should be used for critical sections.

ldts replied:
ok

bjarki-andreasen commented on samples/net/latmon/src/main.c:
```diff
@@ -0,0 +1,263 @@
+
+static K_SEM_DEFINE(ack_event, 0, 1);
+
+bool dhcp_done;
```
This shared value should at least be volatile, though optimally be exchanged for an atomic value :)

bjarki-andreasen commented on samples/net/latmon/src/main.c:
```diff
@@ -0,0 +1,263 @@
+		return ret;
+	}
+
+#if defined(LATMON_LOOPBACK_DBG)
```
is this a leftover? if its still needed it should be a kconfig :)

bjarki-andreasen reviewed (changes requested):
Really nice work! Added a few new notes :)

ldts commented:
AFAICS all actions have been done.

bjarki-andreasen reviewed (dismissed):
Nice work!

bjarki-andreasen commented:
ping @kartben :)

ldts commented:
@kartben ping

kartben commented:
Sorry for the delay. I will make sure to come back to this within a couple days

ldts commented:
> Sorry for the delay. I will make sure to come back to this within a couple days 

Cool thanks, yeah enough for me to know is being reviewed and moving along, didn't want it to stall.

kartben commented on samples/net/latmon/README.rst:
```diff
@@ -0,0 +1,190 @@
+it.
+
+The SUT must be running Latmus to capture the latency metrics and histogram information reported over the
+network. The program uses LEDS to indicate the different states, such as DHCP binding(red), waiting for the
```
```suggestion
network. The program uses LEDs to indicate the different states, such as DHCP binding(red), waiting for the
```

kartben commented on samples/net/latmon/README.rst:
```diff
@@ -0,0 +1,190 @@
+
+- **Connect GPIO pins for transmit (Zephyr to SUT) and receive (SUT to Zephyr)**
+
+  On **FRMD-K64F**, the sample code uses the **Arduino header J2**, ``pin 20`` for transmit the pulse to
```
```suggestion
  On **FRDM-K64F**, the sample code uses the **Arduino header J2**, ``pin 20`` for transmit the pulse to
```

kartben commented on samples/net/latmon/README.rst:
```diff
@@ -0,0 +1,190 @@
+Example
+=======
+
+On the host and to build and flash the Zephyr FRMD-K6F board with the Latmon sample:
```
```suggestion
On the host and to build and flash the Zephyr FRDM-K64F board with the Latmon sample:
```

kartben commented on samples/net/latmon/README.rst:
```diff
@@ -0,0 +1,190 @@
+
+The Latmon sample application is divided into two main components:
+
+- **Application Logic** (``samples/net/latmon/main.c``):
```
```suggestion
- **Application Logic** (:zephyr_file:`samples/net/latmon/main.c`):
```

kartben commented on samples/net/latmon/README.rst:
```diff
@@ -0,0 +1,190 @@
+   It initializes networking, provides the instrumentation mechanism and handles LED indicators for the
+   different states.
+
+- **Library** (``subsys/net/lib/latmon/latmon.c``):
```
```suggestion
- **Library** (:zephyr_file:`subsys/net/lib/latmon/latmon.c`):
```

kartben commented on subsys/net/lib/latmon/latmon.c:
```diff
@@ -0,0 +1,568 @@
+static int send_sample_data(int latmus, struct latmon_data *data)
+{
+	struct latmon_net_data ndata = {
+		.sum_lat_lo = htonl(data->sum_lat & 0xffffffff),
+		.sum_lat_hi = htonl(data->sum_lat >> 32),
```
can't `htonll` be used to simplify this?

ldts replied:
this is the interface with latmus (ie: https://source.denx.de/Xenomai/xenomai4/libevl/-/blob/master/latmus/gpio.c?ref_type=heads#L274 ) . We could change it but then it would affect previous users for not much gain.

kartben commented on doc/connectivity/networking/api/latmon.rst:
```diff
@@ -0,0 +1,151 @@
+        int server_socket, client_socket;
+
+        /* Create and configure the server socket */
+        server_socket = net_latmon_get_socket();
```
```suggestion
        server_socket = net_latmon_get_socket(NULL);
```

kartben commented on include/zephyr/net/latmon.h:
```diff
@@ -0,0 +1,97 @@
+ * @details This function creates and returns a socket to wait for Latmus
+ * connections
+ *
+ * @param ip The IP address to bind the socket to. If NULL, the socket
+ * will be bound to the first available address using the build time configured
+ * latmus port.
+ *
+ * @return A valid socket descriptor on success, negative errno code on failure.
+ */
+int net_latmon_get_socket(struct sockaddr *ip);
```
naming the parameter `ip` is a bit confusing

kartben commented on samples/net/latmon/README.rst:
```diff
@@ -0,0 +1,190 @@
+
+.. code-block:: console
+
+   host$ west build -b frdm_k64f samples/net/latmon
```
more a nit than anything but if you would use a shell prompt like e.g. `user@host:~$ ` then it will be possible for the user to copy paste commands from the docs using the "copy" button with the prompts automatically being discarded. Apply everywhere you have console blocks

ldts commented on doc/connectivity/networking/api/latmon.rst:
```diff
@@ -0,0 +1,151 @@
+
+The following options may be configured to customize the Latmon service:
+
+- :kconfig:option:`CONFIG_NET_LATMON_PORT` - Port number for the Latmon service.
```
um not really, I wasn't planning on doing it. I only commented on that one since it is an agreement with an external subsystem (a separate board). All the others are just the usual Zephyr ones. I hope this is ok.

kartben commented:
> @kartben when I accepted your recommendations, a Co-author tag was generated. Do these need to go into the two commits or shall I just add your Reviewed-by tags?
> 
> going to fold the commits and update the PR so please let me know

Oh I didn't know it does that. No tag at all is fine as far as I'm concerned :) thanks for checking!

ldts committed:
samples: net: latmon: latency monitor tool

Supports the EVL/Xenomai4 benchmarking tool as described in the
documentation. The benchmarking tool is accessed via the latmon service.

This code uses the J2 socket on FRDMk64-F:
             PIN_20: tx pulse to SUT
             PIN_18: rx ack from SUT

The client code running on the SUT shall monitor the falling edge of
PIN_20.

Example usage from the latmus client running Xenomai4:
 $ latmus -I gpiochip2,23,falling-edge  \
          -O gpiochip2,21 -g"histogram" \
          -z broadcast

Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>

ldts commented:
@kartben any other changes?

ldts commented on doc/connectivity/networking/api/latmon.rst:
```diff
@@ -0,0 +1,151 @@
+The service provides functions to create, bind, and listen on sockets.
+It also supports sending and receiving data over TCP and UDP.
+
+Message Queue
+=============
+
+A message queue (``latmon_msgq``) is used to pass messages between threads,
+ensuring thread-safe communication.
+
+Synchronization
+===============
+
+Semaphores (``latmon_done`` and ``monitor_done``) are used to synchronize the monitoring
+process and ensure proper cleanup.
```
sorry didnt notice this. ok doing it now and pushing the update.
